core/src/main/java/org/apache/calcite/rel/metadata/RelMdMeasure.java

```diff
@@ -0,0 +1,214 @@
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.rex.RexSubQuery;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.fun.SqlStdOperatorTable;
+import org.apache.calcite.sql.type.SqlTypeUtil;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static com.google.common.collect.Iterables.getOnlyElement;
+
+/**
+ * Default implementations of the
+ * {@link org.apache.calcite.rel.metadata.BuiltInMetadata.Measure}
+ * metadata provider for the standard logical algebra.
+ *
+ * @see org.apache.calcite.rel.metadata.RelMetadataQuery#isMeasure
+ * @see org.apache.calcite.rel.metadata.RelMetadataQuery#expand
+ */
+public class RelMdMeasure
+    implements MetadataHandler<BuiltInMetadata.Measure> {
+  public static final RelMetadataProvider SOURCE = ReflectiveRelMetadataProvider
+      .reflectiveSource(new RelMdMeasure(), BuiltInMetadata.Measure.Handler.class);
+
+  @Override public MetadataDef<BuiltInMetadata.Measure> getDef() {
+    return BuiltInMetadata.Measure.DEF;
+  }
+
+  /** Catch-all implementation for
+   * {@link BuiltInMetadata.Measure#isMeasure(int)},
+   * invoked using reflection.
+   */
+  public Boolean isMeasure(RelNode rel, RelMetadataQuery mq, int column) {
+    return false;
+  }
+
+  /** Catch-all implementation for
+   * {@link BuiltInMetadata.Measure#expand(int, BuiltInMetadata.Measure.Context)},
+   * invoked using reflection.
+   */
+  public RexNode expand(RelNode rel, RelMetadataQuery mq, int column,
+      BuiltInMetadata.Measure.Context context) {
+    throw new UnsupportedOperationException("expand(" + rel + ", " + column
+        + ", " + context);
+  }
+
+  /** Refines {@code expand} for {@link RelSubset}; called via reflection. */
+  public RexNode expand(RelSubset subset, RelMetadataQuery mq, int column,
+      BuiltInMetadata.Measure.Context context) {
+    for (RelNode rel : subset.getRels()) {
+      // Does not loop. We assume that every RelNode can expand.
+      return mq.expand(rel, column, context);
+    }
+    return expand((RelNode) subset, mq, column, context);
+  }
+
+  /** Refines {@code expand} for {@link Filter}; called via reflection. */
+  public RexNode expand(Filter filter, RelMetadataQuery mq,
+      int column, BuiltInMetadata.Measure.Context context) {
+    return mq.expand(filter.getInput().stripped(), column, context);
+  }
+
+  /** Refines {@code expand} for {@link Project}; called via reflection. */
+  public @Nullable RexNode expand(Project project, RelMetadataQuery mq,
+      int column, BuiltInMetadata.Measure.Context context) {
+    final RexNode e = project.getProjects().get(column);
+    final BuiltInMetadata.Measure.Context context2 =
+        Contexts.forProject(project, context);
+    switch (e.getKind()) {
+    default:
```

Comment:
what's the motivation for putting the default branch first?

Reply:
+1

Reply:
I don't recall. Fixed.